[transactional-tests] Initial runner implementation

tnowacki opened the pull request:
- New test runner that uses struct opt and functional-test like command sections/blocks
- Apologies for the lack of documentation currently. There was more in internal docs that I will try to share at some point, just trying to move this forward ASAP to help testing efforts 

## Motivation

- Will serve for a simpler replacement of the functional-tests and e2e tests
- Lots of tiny things to fix, but that can be done as this is being used

## Test Plan

- Added example tests

vgao1996 commented on language/testing-infra/transactional-test-runner/src/tasks.rs:
```diff
@@ -0,0 +1,319 @@
+}
+
+#[derive(Debug, StructOpt)]
+pub enum ProvidedCommand {
```
I've been trying implement a DiemTestAdapter in the past few days, and based on that experience, we might eventually need the ability to specify additional options for the provided commands. 

This is to set certain parameters for Diem transactions. Here's some code copied from `functional-tests` to give you an idea what these parameters are.
```

struct TransactionParameters<'a> {
    pub sender_addr: AccountAddress,
    pub secondary_signer_addresses: Vec<AccountAddress>,
    pub pubkey: &'a Ed25519PublicKey,
    pub privkey: &'a Ed25519PrivateKey,
    pub secondary_pubkeys: Vec<Ed25519PublicKey>,
    pub secondary_privkeys: Vec<&'a Ed25519PrivateKey>,
    pub sequence_number: u64,
    pub max_gas_amount: u64,
    pub gas_unit_price: u64,
    pub gas_currency_code: String,
    pub expiration_timestamp_secs: u64,
}
```

tnowacki replied:
I think we should somehow look at being able to provide additional commands to these in the trait. We don't want to pollute these for the VM side of things.

vgao1996 replied:
> I think we should somehow look at being able to provide additional commands to these in the trait.

Hmm... do you mean adding a command to set these parameters, which will be applied to all (or maybe next) subsequent transactions? I think that could work! Although depending on the actual design, it might get a little bit messy in certain cases, when you only want these to affect 1 transaction.

> We don't want to pollute these for the VM side of things.

In general I agree, and I certainly don't think we should just add these parameters to the Run command. What I was thinking about was adding an optional string field that will be further parsed by the adapter. IMO the reason why the provided commands even exists is because they are common operations across different platforms, but I think we do have to recognize that many of these platforms will want some level of customizability.

tnowacki replied:
I'm saying you can do something of like `ProvidedCommand<ExtraArgs>` and then populate that in the macro, and tie together with an associated type in the trait. Won't be pretty but... it might work.... maybe

Or you just gut it and share less code, that works too and shouldn't be that big of deal

vgao1996 replied:
> I'm saying you can do something of like ProvidedCommand<ExtraArgs> and then populate that in the macro, and tie together with an associated type in the trait. Won't be pretty but... it might work.... maybe

Good to hear that you are okay with this... at least to some extent...😆

Anyways, this adds another option we could consider. I likely won't (and shouldn't) include this in my first PR so let's not worry too much about it right now. It might become clearer to us which solution would work at that time...

tnowacki commented on language/testing-infra/transactional-test-runner/src/vm_test_harness.rs:
```diff
@@ -0,0 +1,228 @@
+    fn init(pre_compiled_deps: Option<&'a FullyCompiledProgram>) -> Self {
+        Self {
+            compiled_state: CompiledState::new(pre_compiled_deps),
+            default_syntax: None,
```
Yeah I think `init` will be necessary for Genesis in the diem side of things, but I could be wrong. regardless not sure bout the `new` comment, basically the same as init with a different name

tnowacki replied:
But I'll take a look at cleaning up the default syntax thing

vgao1996 replied:
> init will be necessary for Genesis in the diem side of things

This didn't seem to be the case to me. `fn run_test_impl` is the entry point of the harness, and it takes an adapter that's already initialized. It would have made sense to me to have `fn init` being part of the trait if it is called by the framework in some way, which it isn't. It's really weird that the trait is telling me how I should implement a function I use internally. I can definitely see adapters which doesn't want the `Option<&'a FullyCompiledProgram>` or wants to have other parameters in the signature of its constructor.

tnowacki replied:
> I can definitely see adapters which doesn't want to take an Option<&'a FullyCompiledProgram> or wants to have other parameters in the signature of its constructor.

What adapters are these? we are building a testing framework for just the VM and Diem, not the future of Move

vgao1996 replied:
> I can definitely see adapters which doesn't want to take an Option<&'a FullyCompiledProgram> or wants to have other parameters in the signature of its constructor.

I guess it's better to say this is just the surface of the problem, or one particular way it is exemplified. The key of my argument here is that, the adapter trait shouldn't really care what my constructor looks like, or whether it exists at all. By the time I hand over control to your framework, I pass you an adapter that's already created and it doesn't matter how I created it. It's not like the framework needs to create new adapters on the fly. Therefore I think we should just delete this function from the trait.

tnowacki commented on language/testing-infra/transactional-test-runner/src/lib.rs:
```diff
@@ -0,0 +1,8 @@
+
+#![forbid(unsafe_code)]
+
+pub mod framework;
```
executor would imply its just processing Move programs to me. The module has some other code that it is not included in the  "execution" of code, it is really the bulk of the testing framework (besides the parsing of tasks), hence the name

vgao1996 commented on language/testing-infra/transactional-test-runner/src/tasks.rs:
```diff
@@ -0,0 +1,319 @@
+        #[structopt(long = "syntax")]
+        syntax: Option<SyntaxChoice>,
+        #[structopt(long = "address", parse(try_from_str = parse_account_address))]
+        address: Option<AccountAddress>,
```
We may need to change this to a string to support named addresses, or make the address an associated type so that the adapter can decide what to do with it.

tnowacki replied:
Not sure if I follow

vgao1996 replied:
Just wondering if we'll ever want to use named addresses here in the future. (Totally fine with it being a hex address for now.)

tnowacki replied:
Yup, we likely will need to

vgao1996 commented on language/testing-infra/transactional-test-runner/tests/vm_test_harness/example.move:
```diff
@@ -0,0 +1,57 @@
+}
+}
+
+//# run --signers 0x1
+
+script {
+    fun main(s: signer) {
+        0x42::N::give(&s)
+    }
+}
+
+//# view
+//#     --address 0x1
+//#     --resource 0x42::N::R<u64>
+
+//# run --signers 0x1 --syntax=mvir
```
Can you add an example of calling a function?

vgao1996 commented on language/testing-infra/transactional-test-runner/src/tasks.rs:
```diff
@@ -0,0 +1,319 @@
+    },
+}
+
+#[macro_export]
+macro_rules! define_commands {
+    ($command:ident $(, $subcommand:ident)?) => {
+        #[allow(unused_imports)]
+        use structopt::*;
+        #[derive(Debug, StructOpt)]
+        #[structopt(
+            name = "task",
+            about = "Transactional task commands",
+            rename_all = "kebab-case"
+        )]
+        pub enum $command {
+            #[structopt(flatten)]
+            Normal($crate::tasks::ProvidedCommand),
+
+            $(
+                #[structopt(flatten)]
+                Subcommand($subcommand),
+            )?
+        }
+    }
+
+}
```
Could I ask why we are defining this macro instead of using generics? Is it because of structopt? 
```
enum Command<S> {
    Provided(ProvidedCommand),
    Subcommand(S),
}
```

From a trait implementor's point of view, it is very confusing that I need to define an enum of (ProvidedCommand + Subcommand) and implement a split function, given that I've already specified the associated Subcommand type. IMO this should be done by the framework.

tnowacki replied:
StructOpt can't handle generics so we are stuck doing some wonky stuff

vgao1996 replied:
It's also quite confusing to see that for the vm test harness,`EmptySubcommand` is not part of the enum definition. I mean, I get it, it's an empty command and we don't really need to process any instance of it, but it still bothers me that we are defining the associated type, but then see that there's only one enum variant to check, and yet the match is exhaustive:
```
run_test_impl(path, &mut adaptor, |c| match c {
    TaskCommand::Normal(p) => Either::Left(p),
})
```

vgao1996 replied:
> StructOpt can't handle generics so we are stuck doing some wonky stuff

Let me see if it's possible to implement the trait manually without automatic derivation.

tnowacki replied:
> I still think that if possible, we shouldn't require the client to implement a split function, which is a leaky abstraction.

I don't worry too too much about leaky abstractions here. We are building a very simple testing framework. Things can be a tad bit leaky

vgao1996 commented on language/testing-infra/transactional-test-runner/src/framework.rs:
```diff
@@ -0,0 +1,402 @@
+    path: &Path,
+    adaptor: &mut Adaptor,
+    split_command: Split,
+) -> Result<(), Box<dyn std::error::Error>>
```
Why not just use `anayhow::Result` here?

tnowacki replied:
Tried changing this, and remembered it is so datatest_stable::harness! was happy

tnowacki commented:
@vgao1996 some updates here:
- Added support for extra arguments on publish and run
- Changed how subcommands work  (might change this again to deal as I type this out)

For most of your outstanding comments about leakiness or generalness: We are building a testing framework for the VM and the Diem adapter, and it would be incorrect to think about any other usecase in the future. The name of the game here is mostly code resuse and not so much bullet-proof abstractions. I'd rather just copy-paste the code to solve reuse between the VM and Diem adapter than build some crazy, ultra-general framework. But I think we are in a decent middle ground with some macros

tnowacki commented:
Changed subcommands again

vgao1996 commented on language/testing-infra/transactional-test-runner/src/tasks.rs:
```diff
@@ -0,0 +1,317 @@
+        )]
+        pub enum $command {
+            #[structopt(name = "publish")]
+            Publish($crate::tasks::PublishCommand $(, $expub:ident)?),
```
This doesn't seem to work. First, the `:ident` part should be removed. Second, after removing `:ident`, I got the following error when trying to add an extra argument:
```
error: non single-typed tuple enums are not supported
   --> language/testing-infra/transactional-test-runner/src/tasks.rs:245:13
    |
245 | /             #[structopt(name = "publish")]
246 | |             Publish($crate::tasks::PublishCommand $(, $expub)?),
    | |_______________________________________________________________^
    | 
   ::: language/testing-infra/transactional-test-runner/src/vm_test_harness.rs:27:1
    |
27  |   define_commands!(TaskCommand, publish: u64);
    |   -------------------------------------------- in this macro invocation
```

tnowacki replied:
Good call, fixed

vgao1996 commented:
> than build some crazy, ultra-general framework.

@tnowacki you might got my message wrong. To be clear, overall I'm pretty happy with the current design and I'm now fairly confidant that the adapter/command model will work for Diem (with some minor tweaks). I have no intention to complicate it drastically. (Although I do believe that the framework, *in its current form*, is already general enough to handle more clients.)

I'm seeing the leaky abstraction and the inclusion of `fn init` as relatively minor and isolated problems. I'm mentioning them because I think there exists drop-in replacements that still achieve your goal of code reusing, but don't suffer from the issues I raised. I think it's fine we deal with them later.

How about we get this landed first so you can start migrating some of the tests? I can do the fixes in my PR later.

tnowacki commented:
@vgao1996, any blockers remaining on this for you? If not, it would be sweet to start migrating things

tnowacki commented:
> I'm seeing the leaky abstraction and the inclusion of fn init as relatively minor and isolated problems. I'm mentioning them because I think there exists drop-in replacements that still achieve your goal of code reusing, but don't suffer from the issues I raised. I think it's fine we deal with them later.

I would still argue against it being leaky but its kind of moot. I think we should just get this in and get crackalackin. 
Any major blockers left that I missed though?

vgao1996 commented:
Not really. Let me know when you fix your lint and I'll stamp it then.

vgao1996 commented on language/testing-infra/transactional-test-runner/src/framework.rs:
```diff
@@ -34,19 +42,24 @@ pub struct ProcessedModule {
 
 pub struct CompiledState<'a> {
```
Now with more and more stuff being put into this struct, the name "CompiledState" seems pretty inaccurate. How about something like "AdapterState"?

tnowacki replied:
Maybe... It is still only really compiler information in way. Maybe more like `MoveProgramState` but thats confusing because its state for the program text, not the program executing

vgao1996 commented on language/testing-infra/transactional-test-runner/src/framework.rs:
```diff
@@ -389,41 +419,81 @@ where
-            task_number, task_name, start_line, stop_line, result_string
-        ));
+
+    let first_task = tasks.pop_front().unwrap();
+    let init_opt = match &first_task.command {
+        KnownCommandFormat::Init(_, _) => Some(first_task.map(|known| match known {
+            KnownCommandFormat::Init(command, extra_args) => (command, extra_args),
+            _ => unreachable!(),
+        })),
+        _ => {
+            tasks.push_front(first_task);
+            None
+        }
+    };
```
Nit: I wonder if instead of using a VecDeque and actually moving the in and out of the container, can we just use slices?
```
let non_init_tasks = match tasks.get(0) {
    Some(task) => {
        process_init(task);
        &tasks[1..] 
    }
    None => &tasks[..]
}
```

tnowacki replied:
I need the owned values of the tasks for moving them into `KnownCommandFormat`

vgao1996 reviewed (dismissed):
Overall LGTM

tnowacki commented:
/land

bors-libra commented:
:broken_heart: Test Failed - [ci-test](https://github.com/diem/diem/actions/runs/1117549665)

vgao1996 commented:
/land

bors-libra commented:
:boom: Tests timed-out
